Apps/Patient/src/Services/PatientDataService.cs

```diff
@@ -17,30 +17,43 @@ namespace HealthGateway.Patient.Services
 {
 #pragma warning disable SA1600 // Disable documentation for internal class.
     using System;
+    using System.Collections.Generic;
     using System.Linq;
     using System.Threading;
     using System.Threading.Tasks;
     using AutoMapper;
+    using HealthGateway.AccountDataAccess.Patient;
+    using HealthGateway.Common.Data.Constants;
     using HealthGateway.Common.Services;
+    using HealthGateway.Patient.Constants;
     using HealthGateway.PatientDataAccess;
 
     internal class PatientDataService : IPatientDataService
     {
         private readonly IPatientDataRepository patientDataRepository;
+        private readonly IPatientRepository patientRepository;
         private readonly IPersonalAccountsService personalAccountsService;
         private readonly IMapper mapper;
 
-        public PatientDataService(IPatientDataRepository patientDataRepository, IPersonalAccountsService personalAccountsService, IMapper mapper)
+        public PatientDataService(IPatientDataRepository patientDataRepository, IPatientRepository patientRepository, IPersonalAccountsService personalAccountsService, IMapper mapper)
         {
             this.patientDataRepository = patientDataRepository;
+            this.patientRepository = patientRepository;
             this.personalAccountsService = personalAccountsService;
             this.mapper = mapper;
         }
 
         public async Task<PatientDataResponse> Query(PatientDataQuery query, CancellationToken ct)
         {
+            List<PatientDataType> patientDataTypes = query.PatientDataTypes.ToList();
+
+            if (!await this.CanAccessDataSourceAsync(query.Hdid, patientDataTypes).ConfigureAwait(true))
+            {
+                return new PatientDataResponse(Enumerable.Empty<PatientData>());
+            }
+
```

Comment:
NIT:

This could be simplified to still use the patientRepository's `CanAccessDataSourceAsync` by using it within a linq `Where`. The only caveat which you can use to determine if this is worthwhile or not, would be that you should either create an Async extension for `Where` or force the method into operate syncrhonously using `.Result`.

```CSharp
List<PatientDataType> patientDataTypes = query.PatientDataTypes
                .Where(p => this.patientRepository.CanAccessDataSourceAsync(query.Hdid, this.mapper.Map<DataSource>(p), ct).Result)
                .ToList();

if (!patientDataTypes.Any())
{
    return new PatientDataResponse(Enumerable.Empty<PatientData>());
}
```

You would not need the additional function below.
`private async Task<bool> CanAccessDataSourceAsync(string hdid, IList<PatientDataType> patientDataTypes)`

It does not affect the outcome of the tests:
![image](https://github.com/bcgov/healthgateway/assets/19548348/dc13bb59-ff3a-4800-aa6a-96e2c25caa8c)

Reply:
PatientDataService.CanAccessDataSourceAsync is required as it removes any PatientDataType that have been blocked.

PatientDataTypes is list of PatientDataType so it can have both OrganDonorRegistrationStatus and DiagnosticImaging.  One is blocked while the other is not.  

Line 57 -  PatientDataQueryResult result = await this.patientDataRepository.Query(healthQuery, ct)
                .ConfigureAwait(true);

This queries based on the list.

So if one PatientDataType  from the list is blocked but the other is not, the above code logic will allow it still query.